text/0000-selfexhausting_iter_adapter.md

```diff
@@ -0,0 +1,128 @@
+The current `drain` APIs run the following code on drop before repairing the collection's state:
+
+```rust
+for _ in &mut self {}
+```
+
+That is, they run themselves to the end for the side-effect of calling the destructor on each element to be removed from the collection.
+
+This showcases the use of self-exhausting iterators. The principle can apply to any side effecting iterator where all side effects are needed but not (all) the elements it returns. An `.exhausting()` adapter allows adding self-exhaustion on drop to arbitrary iterators.
+The behaviour of `drain` could be attained by combining two separate functions:
+`drain_nonexhausting().exhausting()`
+
+Iteration through `.by_ref()` and subsequent consumption can achieve the same result, but only when one is holding the iter. With `.exhausting()`, the iter can be passed to a function or returned from one. For this reason, it also works better with method chaining.
+Note that returning a self-exhausting iterator from a function should mostly be limited to callback situations. Hardcoding self-exhaustion is mixing concerns and needlessly limiting.
+Adding `exhausting` to the std library should make it easy for users to gain this behaviour where necessary and avert more non-lazy iterator APIs in the std library and outside of it.
+
+## Examples
+```rust
+// pass side effecting iter away
+iter_of_iters.flat_map(|iter| {
+        iter.map(side_effects)
+            .exhausting() // finish what you've started
+            .take_while(condition)
+    });
+
+// return self-exhausting iter from function
+fn drain(&mut self) -> Drain {
+    // wrapper can forward all iterator methods to internal iter
+    Drain(
+        self.drain_nonexhausting()
+            .exhausting()
+    )
+}
+
+// -------------------- Aesthetic improvements only ---------------------
+// Current: manual exhaustion with by-ref
+let mut iter = iter.some()
+    .adapter()
+    .chain();
+let val = iter.by_ref() // chain breaking indirection
+    .map(func)
+    find(condition);
+iter.for_each(|_| {}); // explicitly consume iterator
+                       // must have access to iter
+
+// With proposal:
+let val = iter.some()   // all of this
+    .adapter()          // will run
+    .chain()            // for all elements in iter
+    .exhausting()
+    .map(func)          // runs only until an element
+    .find(condition);   // is found or iter is exhausted
+//----------------------------------------------------------------------
+```
+
+# Implementation
+
+The `exhausting()` method should take the iterator by value.
+During iteration, `Exhausting` is a trivial wrapper that acts like `&mut Self`, meaning it implements all the Iterator traits that the contained iter implements and will always do external iteration. On drop, it runs `for _ in self {}`.
```

Comment:
Why always external iteration?  What would keep it from doing internal iteration with [`try_fold`](https://doc.rust-lang.org/nightly/std/iter/trait.Iterator.html#method.try_fold)?

Reply:
You're right. I was confused about internal iteration requiring consumption.